### DISCUSSION TITLE
How do `cogs` pass `self` into command callbacks?

### DISCUSSION THREAD

**Discussion by showierdata9978** [nextcord/nextcord#945] [Q&A] [answered]
Normally with function decorators that store the base method, and then the base method is called somewhere else without access to the class, you can't bind `self` to the method (so that it can access insistence vars)

**Comment by EmmmaTech** [comment, marked as answer]
Application commands are bound to a cog when an instance of a cog is created, so when the callback inside of an application command is called, the cog instance is the first argument passed in.

**Reply by showierdata9978** [reply]
so its a more complicated version of this? (just wrote it)

```py

class Cog(metaclass=ABCMeta):
    instence = None
    _Cog_commands = None

    def __new__(cls):
        if cls.instence is None:
            cls.instence = super().__new__(cls)
        return cls.instence

    @classmethod
    def __subclasshook__(cls):
        cls._Cog_commands = {} # to make cogs not share cmds

    @classmethod
    def register_command(cls, command):
        cls._Cog_commands[command.name] = command
    
    
    def get_commands(self):
        for cmd in self._Cog_commands:
            cmd.func = getattr(self, cmd.func.__name__) # binding self 
        
        return self._Cog_commands

... another file


def command(name=None, arg_count=0, bot=None):
    def inner(func):
        @functools.wraps(func)
        def callfunc(*args, **kwargs):
            raise NotImplementedError("This Should Never be called\n it should disapear")
        
        cmd = _Command(func, name, arg_count)

        if hasattr(func, "__class__"):
            func.__class__.register_command(cmd)
        elif bot is not None:
            bot.commands[cmd.name] = cmd
        else:
            Bot.register_command(cmd)


        return callfunc

**Reply by EmmmaTech** [reply]
Not quite. The cog class looks for application commands defined within itself then modifies them to set themselves as the cog in the application command.

https://github.com/nextcord/nextcord/blob/22eb5d169f6a05f2c813d72a158972a0ba3d7dc1/nextcord/application_command.py#L476-L511

**Reply by EmmmaTech** [reply]
The `slash_command`, `user_command`, and `message_command` decorators defined in the `nextcord` module (so not the decorators defined in `nextcord.Client`) just convert the decorated function into an appropriate `BaseApplicationCommand` subclass.
